### ISSUE TITLE
Fine tuning fastspeech-150k.h5 (v3) on ljspeech for 2k more steps results in poor audio quality.

### ISSUE THREAD

**Issue by brendanxwhitaker** [TensorSpeech/TensorFlowTTS#421]
I've downloaded LJSpeech, preprocessed, normalized, extracted durations, and set up for finetuning. I have a dataset for a different voice to do transfer learning, but I was getting poor results, so as a sanity check I tried finetuning on LJSpeech for 2k steps, so total training steps is 152k. This is the result: https://drive.google.com/file/d/1OpcTOx0ozBCVjc8xOHzvPYonyk7I17h5/view?usp=sharing

Result before finetuning:
https://drive.google.com/file/d/1tZxQwVsLiSz2VaGkhOPUXYSYyloPFtuU/view?usp=sharing

I was wondering what might cause this gravely noise? It's much much worse than the 150k model that we started with, which sounds the voice samples linked in the examples page. Durations extracted with `models/tacotron2-model-65000.h5`.

**Comment by dathudeptrai** [comment]
> I've downloaded LJSpeech, preprocessed, normalized, extracted durations, and set up for finetuning. I have a dataset for a different voice to do transfer learning, but I was getting poor results, so as a sanity check I tried finetuning on LJSpeech for 2k steps, so total training steps is 152k. This is the result: https://drive.google.com/file/d/1OpcTOx0ozBCVjc8xOHzvPYonyk7I17h5/view?usp=sharing
> 
> Result before finetuning:
> https://drive.google.com/file/d/1tZxQwVsLiSz2VaGkhOPUXYSYyloPFtuU/view?usp=sharing
> 
> I was wondering what might cause this gravely noise? It's much much worse than the 150k model that we started with, which sounds the voice samples linked in the examples page. Durations extracted with `models/tacotron2-model-65000.h5`.

the problem is caused by the optimizer state. When you re-train with the pretrained 150k fastspeech, the optimizer is reset at `0` step rather than  `150k` steps.

**Comment by brendanxwhitaker** [comment]
What would you recommend to fix this issue? Finetuning for longer so the learning rate drops back down? Or is there a way to preserve or set the optimizer state when you resume?



________________________________

**Comment by brendanxwhitaker** [comment]
Ah okay I didn't see the bit about starting with the smaller learning rate. Thanks! Also, there is a lot of debate on what to set for `var_train_expr:` in https://github.com/TensorSpeech/TensorFlowTTS/issues/296. I was wondering if you could recommend which values to set for that field when finetuning for a different voice?

**Comment by OscarVanL** [comment]
Hi @brendanxwhitaker. The recommendation to fiddle with `var_train_expr` came from the particular problem I was trying to address in that issue, which was a tiny training dataset. In that case, it was to try and avoid overfitting.

In the end, it never led to fruitful results. 

Instead, I finetuned an existing model, while merging my small dataset with a larger dataset, leaving `var_train_expr` as the default value (train everything).

In your case, the problem we were trying to solve in #296 is not relevant. 

Based on my (limited) experience, I'd say leaving `var_train_expr` alone is the best option.

**Comment by brendanxwhitaker** [comment]
@OscarVanL I see, well that's really good to know! I won't mess with it then. Thanks for the advice!

**Comment by OscarVanL** [comment]
The thing that made a huge difference to me was that `hop_size` and `fft_size` did not match between the model I was tuning, my training parameters, and the vocoder I was using. I'd suggest checking these values match.

**Comment by brendanxwhitaker** [comment]
I'm still getting the same issue. I've changed the initial learning rate, as you can see below:

**Comment by brendanxwhitaker** [comment]
The vocoder config:

```
# This is the hyperparameter configuration file for Multi-Band MelGAN.
# Please make sure this is adjusted for the LJSpeech dataset. If you want to
# apply to the other dataset, you might need to carefully change some parameters.
# This configuration performs 1000k iters.

###########################################################
#                FEATURE EXTRACTION SETTING               #
###########################################################
sampling_rate: 22050
hop_size: 256            # Hop size.
format: "npy"


###########################################################
#         GENERATOR NETWORK ARCHITECTURE SETTING          #
###########################################################
model_type: "multiband_melgan_generator"

multiband_melgan_generator_params:
    out_channels: 4               # Number of output channels (number of subbands).
    kernel_size: 7                # Kernel size of initial and final conv layers.
    filters: 384                  # Initial number of channels for conv layers.
    upsample_scales: [8, 4, 2]    # List of Upsampling scales.
    stack_kernel_size: 3          # Kernel size of dilated conv layers in residual stack.
    stacks: 4                     # Number of stacks in a single residual stack module.
    is_weight_norm: false         # Use weight-norm or not.

###########################################################
#       DISCRIMINATOR NETWORK ARCHITECTURE SETTING        #
###########################################################
multiband_melgan_discriminator_params:
    out_channels: 1                   # Number of output channels.
    scales: 3                         # Number of multi-scales.
    downsample_pooling: "AveragePooling1D"   # Pooling type for the input downsampling.
    downsample_pooling_params:        # Parameters of the above pooling function.
        pool_size: 4
        strides: 2
    kernel_sizes: [5, 3]              # List of kernel size.
    filters: 16                       # Number of channels of the initial conv layer.
    max_downsample_filters: 512       # Maximum number of channels of downsampling layers.
    downsample_scales: [4, 4, 4]      # List of downsampling scales.
    nonlinear_activation: "LeakyReLU" # Nonlinear activation function.
    nonlinear_activation_params:      # Parameters of nonlinear activation function.
        alpha: 0.2
    is_weight_norm: false             # Use weight-norm or not.

###########################################################
#                   STFT LOSS SETTING                     #
###########################################################
stft_loss_params:
    fft_lengths: [1024, 2048, 512]  # List of FFT size for STFT-based loss.
    frame_steps: [120, 240, 50]     # List of hop size for STFT-based loss
    frame_lengths: [600, 1200, 240] # List of window length for STFT-based loss.

subband_stft_loss_params:
    fft_lengths: [384, 683, 171]  # List of FFT size for STFT-based loss.
    frame_steps: [30, 60, 10]     # List of hop size for STFT-based loss
    frame_lengths: [150, 300, 60] # List of window length for STFT-based loss.

###########################################################
#               ADVERSARIAL LOSS SETTING                  #
###########################################################
lambda_feat_match: 10.0      # Loss balancing coefficient for feature matching loss
lambda_adv: 2.5              # Loss balancing coefficient for adversarial loss.

###########################################################
#                  DATA LOADER SETTING                    #
###########################################################
batch_size: 64                 # Batch size.
batch_max_steps: 8192          # Length of each audio in batch for training. Make sure dividable by hop_size.
batch_max_steps_valid: 81920   # Length of each audio for validation. Make sure dividable by hope_size.
remove_short_samples: true     # Whether to remove samples the length of which are less than batch_max_steps.
allow_cache: true              # Whether to allow cache in dataset. If true, it requires cpu memory.
is_shuffle: true               # shuffle dataset after each epoch.

###########################################################
#             OPTIMIZER & SCHEDULER SETTING               #
###########################################################
generator_optimizer_params:
    lr_fn: "PiecewiseConstantDecay"
    lr_params:
        boundaries: [100000, 200000, 300000, 400000, 500000, 600000, 700000]
        values: [0.0005, 0.0005, 0.00025, 0.000125, 0.0000625, 0.00003125, 0.000015625, 0.000001]
    amsgrad: false
discriminator_optimizer_params:
    lr_fn: "PiecewiseConstantDecay"
    lr_params:
        boundaries: [100000, 200000, 300000, 400000, 500000]
        values: [0.00025, 0.000125, 0.0000625, 0.00003125, 0.000015625, 0.000001]
    amsgrad: false

###########################################################
#                    INTERVAL SETTING                     #
###########################################################
discriminator_train_start_steps: 200000  # steps begin training discriminator
train_max_steps: 4000000                 # Number of training steps.
save_interval_steps: 20000               # Interval steps to save checkpoint.
eval_interval_steps: 5000                # Interval steps to evaluate the network.
log_interval_steps: 200                  # Interval steps to record the training log.

###########################################################
#                     OTHER SETTING                       #
###########################################################
num_save_intermediate_results: 1  # Number of batch to be saved as intermediate results.
```

**Comment by brendanxwhitaker** [comment]
Command being used to finetune:
```
CUDA_VISIBLE_DEVICES=1 python tensorflowtts/examples/fastspeech/train_fastspeech.py \
  --train-dir ./dump_ljspeech/train/ \
  --dev-dir ./dump_ljspeech/valid/ \
  --outdir ./models/train.ljspeech.fastspeech.v3/ \
  --config ./tensorflowtts/examples/fastspeech/conf/fastspeech.v3.yaml \
  --use-norm 1 \
  --pretrained ./models/fastspeech-150k.h5 \
  --mixed_precision 0 \
  --resume ""
```

**Comment by dathudeptrai** [comment]
@brendanxwhitaker try fastspeech2.

**Comment by brendanxwhitaker** [comment]
@dathudeptrai I have tried that and it fixed the finetuning issue. However, with FS2, I'm getting strange behavior when running inference from C++ (tflite). Please see https://github.com/TensorSpeech/TensorFlowTTS/issues/426

**Comment by PeijiYang** [comment]
> @dathudeptrai I have tried that and it fixed the finetuning issue. However, with FS2, I'm getting strange behavior when running inference from C++ (tflite). Please see #426

@brendanxwhitaker Do you mean that pre-training on fs2 and finetune 2k step can achieve better results?
